src/high-level/matrix.lisp

```diff
@@ -0,0 +1,536 @@
+                      (second shape)
+                      size
+                      (or order :column-major)
+                      (or
+                       storage
+                       (apply #'make-array
+                              size
+                              :element-type ',type
+                              (if initial-element
+                                  (list :initial-element (coerce initial-element ',type))
+                                  nil)))))))
+
+       ;; TODO: This does not allow for args. Make this allow for args.
+       (defmethod copy-tensor ((m ,name) &rest args)
+         (let ((new-m (,copy-sym m)))
+           (setf (,storage-sym new-m)
+                 (make-array (matrix-size m) :element-type (element-type m)))
+           new-m))
+
+       (defmethod deep-copy-tensor ((m ,name) &rest args)
+         (let ((new-m (,copy-sym m)))
+           (setf (,storage-sym new-m)
+                 (alexandria:copy-array (,storage-sym m)))
+           new-m))
+       
+       (defmethod tref ((matrix ,name) &rest pos)
+         (declare (dynamic-extent pos))
+         (let ((numrows (matrix-nrows matrix))
+               (numcols (matrix-ncols matrix)))
+           (declare (type fixnum numrows numcols))
+           (policy-cond:with-expectations
+               (> speed safety)
+               ((assertion (valid-matrix-index-p pos numrows numcols)))
+             (let ((row (first pos))
+                   (col (second pos)))
+               (declare (type fixnum row col))
+               (let ((index (ecase (matrix-order matrix)
+                              (:row-major (cl:+ col (the fixnum (* row numcols))))
+                              (:column-major (cl:+ row (the fixnum (* col numrows)))))))
+                 (declare (type alexandria:array-index index))
+                 (aref (,storage-sym matrix) index))))))
+       
+       (defmethod (setf tref) (new-value (matrix ,name) &rest pos)
+         (declare (dynamic-extent pos))
+         (let ((numrows (matrix-nrows matrix))
+               (numcols (matrix-ncols matrix)))
+           (declare (type alexandria:non-negative-fixnum numrows numcols))
+           (policy-cond:with-expectations
+               (> speed safety)
+               ((assertion (valid-matrix-index-p pos numrows numcols)))
+             (let ((row (first pos))
+                   (col (second pos)))
+               (declare (type alexandria:non-negative-fixnum row col))
+               (let ((index (ecase (matrix-order matrix)
+                              (:row-major (cl:+ col (the fixnum (* row numcols))))
+                              (:column-major (cl:+ row (the fixnum (* col numrows)))))))
+                 (declare (type alexandria:array-index index))
+                 (setf (aref (,storage-sym matrix) index)
+                       (coerce new-value ',type))))))))))
+
+(defun pprint-matrix (stream matrix &optional colon-p at-sign-p)
```

Comment:
For integer matrices, this (and similarly `pprint-vector`) prints values as if they were floats, e.g.
`(magicl:from-list '(1 1) '(2))` prints as
```
#<MAGICL::VECTOR/INT32 (2):
   1.000
   1.000>
```
this is harmless, but it did make me do a double take.

Reply:
Fixed 4ed8244